### PR TITLE
Use multiple gradients on a single ingredient

### LINKED ISSUE #262: Combine multiple gradients on a single ingredient

### PR DESCRIPTION
**Pull request by mogres**
Problem
=======
Closes #262 

Solution
========
Added support to allow multiple gradients on a single ingredient at the ingredient level.

## Type of change
* New feature (non-breaking change which adds functionality)

Change summary:
---------------
* Allow the `gradient` attribute of an ingredient to be a `List` instance with multiple gradients.
* While creating an ingredient, check if it requires multiple gradients. If yes, add an attribute `combined_weight` that specifies the probabilities of the grid point to be picked to place the ingredient.
* Refactor functions in the `Gradient` class to be static methods.
* Add example recipe and edit config.

Steps to Verify:
----------------
Pack a recipe with an X and Y gradient combined:
```bash
pack -r cellpack/tests/recipes/v2/test_combined_gradient.json -c cellpack/tests/packing-configs/test_grid_plot_config.json
```

Screenshots:
-----------------------

#### Weight map:
combined_gradient

#### Simularium screenshot
image

### CODE REVIEW

File: cellpack/autopack/Analysis.py [hunk 1]
```diff
@@ -828,9 +828,12 @@ def update_distance_distribution_dictionaries(
...
         if ingr.packing_mode == "gradient" and self.env.use_gradient:
-            self.center = center = self.env.gradients[ingr.gradient].mode_settings.get(
-                "center", center
-            )
+            if not isinstance(ingr.gradient, list):
+                self.center = center = self.env.gradients[
+                    ingr.gradient
+                ].mode_settings.get("center", center)
+            else:
+                self.center = center
```
Comment (rugeli) [inline cellpack/autopack/Analysis.py:836]: nit: maybe simply reverse `if` and `else` statements for readability?
```suggestion
            if isinstance(ingr.gradient, list):
                self.center = center
            else:
                self.center = self.env.gradients[
                    ingr.gradient
                ].mode_settings.get("center", center)
```

**Review by rugeli — APPROVED** [review]
I was able to run and see the combined gradient results, very cool! Just one minor suggestion, everything else look good

File: cellpack/autopack/Environment.py [hunk 2]
```diff
@@ -1700,7 +1700,24 @@ def getPointToDrop(
...
                 self.log.info("pick point from gradients %d", (len(allIngrPts)))
-                ptInd = self.gradients[ingr.gradient].pickPoint(allIngrPts)
+                if isinstance(ingr.gradient, list) and len(ingr.gradient) > 1:
+                    if not hasattr(ingr, "combined_weight"):
+                        gradient_list = [
+                            gradient
+                            for gradient_name, gradient in self.gradients.items()
+                            if gradient_name in ingr.gradient
+                        ]
+                        combined_weight = Gradient.get_combined_gradient_weight(
```
Comment (meganrm) [inline cellpack/autopack/Environment.py:1710]: can you push the logic above into a generic Gradient.pick_point_from_weight that figures out wether it needs to combine weights and returns the ptInd

File: cellpack/autopack/Analysis.py (hunk 1 above)
Comment (mogres) [reply cellpack/autopack/Analysis.py]: Thanks Ruge! This reminded me of another edge case of a single gradient in a list so I added that in as well [here](https://github.com/mesoscope/cellpack/pull/270/commits/5614cd0285d7bd401f0c94d0a56efd73a0bf1265).

File: cellpack/autopack/Environment.py (hunk 2 above)
Comment (mogres) [reply cellpack/autopack/Environment.py]: Good call! I added a static method `Gradient.pick_point_for_ingredient` [here](https://github.com/mesoscope/cellpack/pull/270/commits/6cb42589eeda1fd88b6bd5b90d6a8039bdd99181)

**Review by meganrm — APPROVED** [review]

**Review by rugeli — APPROVED** [review]

**Pull request merged**